### DISCUSSION TITLE
Double-weighted directed graph

### DISCUSSION THREAD

**Discussion by kyper-jeff** [networkx/networkx#6011] [Q&A]
I have a graph theoretical problem, which is unfortunate for me given I know nothing about graph theory. Apologies in advance for butchering the terminology.

Suppose I have a directed graph (with edges (A,B) and (B,A) both possible and distinct from each other), with vertices and edges assigned positive integer-valued weights. The vertex weights will always be greater than or equal to the sum of the edge weights that connect them to other vertices.

My first question concerns how such a graph would be represented in an adjacency list. I've seen weighted graphs represented in adjacency lists with the connected vertices and their edge weights noted, but I've not encountered a way to represent both edge weights and vertex weights in an adjacency list. How would I do this in networkx?

**Comment by dschult** [comment]
The edge weights are represented by a matrix (two nodes are needed to specify the location of the value in the matrix).
Node attributes are represented by a vector (one node is needed to specify the location of the value in the vector).
But you can combine them in creative ways. For example, you could create an edge attribute that takes the edge "weight" and subtracts the node values on each end of that edge. This would be an edge attribute, so you can make an adjacency matrix with that attribute and it contains the information about both edge weights and node weights in its formula.

**Comment by kyper-jeff** [comment]
Hi Dan

Thanks very much for your email, and for walking me through the flexibility of networkx graph representations.

Without wanting to take advantage of your generosity, may I ask how some of the algorithms would use the vertex weights encoded in the weight vector? Would I always need to transform the edge weights in the way you’ve described (since the graph clustering or path search algorithms rely on edge weights only), or do some of the algorithms make it possible to consider both edge and vertex weights separately?

Kind regards,

Jeff Stephenson

**Reply by rossbar** [reply]
> how some of the algorithms would use the vertex weights encoded in the weight vector?

The vast majority of functions in networkx expect the graph to be represented as NetworkX graph objects (`nx.Graph`, `nx.DiGraph`, `nx.MultiGraph`, etc.). If you are planning on using NetworkX to analyze graph data, it is very likely you will have to convert from e.g. the adjacency matrix/ node attribute list into a NetworkX graph.

> do some of the algorithms make it possible to consider both edge and vertex weights separately?

This depends on the specific algorithm - some may have keyword arguments that allow you to specify the attribute names for node/edge attributes. Some algorithms allow you to specify a function via the `weight` parameter so that you can implement custom weights that account for both vertex and edge weights. The [Notes section of the `dijkstra_path`](https://networkx.org/documentation/latest/reference/algorithms/generated/networkx.algorithms.shortest_paths.weighted.dijkstra_path.html) has a nice example of how one might do this.
